In [1]:
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np
import math 
%load_ext sql

import matplotlib.pyplot as plt 
%matplotlib inline

### Query the AWS database for raw data table using 'ipython-sql'

In [2]:
def get_top_3_item_from_db(top10only):
    aItems0 = str(top10only.ITEM.values[0])
    aItems1 = str(top10only.ITEM.values[1])
    aItems2 = str(top10only.ITEM.values[2])
    
    %sql postgresql://sharknado:sharknado@sharknado-dse.ceg3hdkdq8l0.us-east-1.rds.amazonaws.com/sharknado
    result = %sql select distinct item,vend,l2,l3 from products \
    where item in (:aItems0,:aItems1,:aItems2)
     
    local_df = result.DataFrame()
    
    return local_df

## 1. What are the most popular item(s) for a chosen product/category?

### Popular items for salty-snack in 2008 to 2011

In [3]:
dir_path = "/media/mqh001/Data/DSE220_Project_Iris_dataset/IRITest/"

#find saltsnck_drug *
file1 = dir_path + 'Year8/saltsnck/saltsnck_drug_1479_1530'
file2 = dir_path + 'Year9/saltsnck/saltsnck_drug_1531_1582'
file3 = dir_path + 'Year10/saltsnck/saltsnck_drug_1583_1634'
file4 = dir_path + 'Year11/saltsnck/saltsnck_drug_1635_1686'

fileList = [file1, file2, file3, file4]
fileList_size = len(fileList)

In [4]:
top3saler4years = pd.DataFrame()
store_df = pd.DataFrame()

for i, aFile in zip(range(1,fileList_size+1),fileList):
    aYear_df = pd.read_fwf(aFile)
    print "Year%d - panel_dr.shape ="%(i+7), aYear_df.shape
    print "   Number of items in saltysnack = ", len(aYear_df['ITEM'].unique())
    print "   Number of stores = ", len(aYear_df['IRI_KEY'].unique())
    
    aYear_df = aYear_df[aYear_df.UNITS != 0]
    
    #To avoid outliner will ruin my data, I calculate the average price for each sale transaction first.
    #Set every price to be an unit value price.
    aYear_df['DOLLARS_2'] = aYear_df['DOLLARS']/aYear_df['UNITS']
    aYear_df['UNITS_2'] = 1
    aYear_df = aYear_df.join(aYear_df.groupby(['ITEM'])['DOLLARS_2'].sum(), on='ITEM', rsuffix='_SUM')
    aYear_df = aYear_df.join(aYear_df.groupby(['ITEM'])['UNITS_2'].sum(), on='ITEM', rsuffix='_SUM')
    aYear_df['DOLLARS_AVG'] = aYear_df['DOLLARS_2_SUM']/aYear_df['UNITS_2_SUM']
    
    #Calculate the total sale price and units
    aYear_df = aYear_df.join(aYear_df.groupby(['ITEM'])['DOLLARS'].sum(), on='ITEM', rsuffix='_SUM')
    aYear_df = aYear_df.join(aYear_df.groupby(['ITEM'])['UNITS'].sum(), on='ITEM', rsuffix='_SUM')
      
    total_items = aYear_df['UNITS'].sum()
    total_dollars = aYear_df['DOLLARS'].sum()
    
    #print "   Total null values:\n", aYear_df.isnull().sum(), '\n\n'
    store_df = pd.concat([store_df,aYear_df])
    
    #Get the top 3 popular items
    items = aYear_df.groupby(['ITEM'])['UNITS'].sum() 
    
    items_df = pd.DataFrame(items)
    items_df['TOTAL_SALE'] = aYear_df.groupby('ITEM')['DOLLARS_SUM'].sum()
    
    items_df = items_df.sort(['UNITS'], ascending=[0])
    items_df = items_df.reset_index()
 
    top3saler = items_df.head(3)
    top3saler['PER'] = top3saler['UNITS']/total_items
    top3saler['PER'] = top3saler['PER'].apply(lambda x: '%.2f'%(x*100)).astype(float)
    
    #top3saler = items_df.head(3)
    df = get_top_3_item_from_db(top3saler)

    df = df.groupby(['item']).agg(lambda col: '    '.join(col))[['vend', 'l2','l3']]
    df['year'] = 2007+i
    df['ave_cost'] = 0
    df['ave_cost'][:] = aYear_df['DOLLARS_AVG'][aYear_df['ITEM'][top3saler['ITEM']]]
    df['sale_%'] = 0
    df['sale_%'][:] = top3saler['PER']  
    #Put top10saler to 4 years table
    top3saler4years = pd.concat([top3saler4years,df])

top3saler4years['ave_cost'] = top3saler4years['ave_cost'].apply(lambda x: '%.2f'%x)

print "top3saler4years.shape =", top3saler4years.shape

Year8 - panel_dr.shape = (1641963, 11)
   Number of items in saltysnack =  2695
   Number of stores =  523
3 rows affected.
Year9 - panel_dr.shape = (1635324, 11)
   Number of items in saltysnack =  2672
   Number of stores =  504
3 rows affected.
Year10 - panel_dr.shape = (1727963, 11)
   Number of items in saltysnack =  2529
   Number of stores =  484
3 rows affected.
Year11 - panel_dr.shape = (1801550, 11)
   Number of items in saltysnack =  2609
   Number of stores =  463
3 rows affected.
top3saler4years.shape = (12, 6)


In [5]:
top3saler4years.to_csv('top3saler4years_2008_11.csv', encoding='utf-8')

In [6]:
top3saler4years = pd.read_csv('top3saler4years_2008_11.csv')

In [32]:
top3saler4years2 = top3saler4years
top3saler4years2.columns=['item', 'vendor', 'category', 'parent company', 'year', 'cost per unit ($)', 'sale percentage (%)']

### The top 3 popular items from 2008 to 2011

In [33]:
top3saler4years2.reset_index().set_index('year')

index   item  vendor                category    parent company  \
year                                                                   
2008      0  37705   37000            POTATO CHIPS  PROCTER & GAMBLE   
2008      1  37717   37000            POTATO CHIPS  PROCTER & GAMBLE   
2008      2   6452   28400  TORTILLA/TOSTADA CHIPS       PEPSICO INC   
2009      3   7858   28400            POTATO CHIPS       PEPSICO INC   
2009      4   7881   28400  TORTILLA/TOSTADA CHIPS       PEPSICO INC   
2009      5   8321   28400            POTATO CHIPS       PEPSICO INC   
2010      6   7858   28400            POTATO CHIPS       PEPSICO INC   
2010      7   7881   28400  TORTILLA/TOSTADA CHIPS       PEPSICO INC   
2010      8   8321   28400            POTATO CHIPS       PEPSICO INC   
2011      9   7858   28400            POTATO CHIPS       PEPSICO INC   
2011     10   7874   28400           CHEESE SNACKS       PEPSICO INC   
2011     11   7881   28400  TORTILLA/TOSTADA CHIPS       PEPSICO INC   

      cost per unit ($)  sale percentage (%)  
year                                          
2008               3.49                 2.84  
2008               1.44                 2.05  
2008               3.50                 1.92  
2009               0.98                 2.08  
2009               0.98                 2.07  
2009               1.25                 1.98  
2010               0.98                 2.39  
2010               0.97                 2.03  
2010               2.00                 2.02  
2011               3.60                 1.99  
2011               1.27                 1.87  
2011               3.46                 1.60

### Analysis: 
- POTATO CHIPS has been the most popular salty-snacks since 2009 to 2011.  And the top 3 popular items supplier comes from one parent company 'PEPSICO INC'.

## 2. How does the popularity of an item change with the cost associated with it for a chosen year?

Answer: 
- Item 7858-POTATO CHIPS and 7881-TORTILLA/TOSTADA CHIPS are the most popular.  However, 7858-POTATO CHIPS is a better saler. 
- During these 4 years, Cheese Snacks only has been made in the top 3 list once in 2011. And it could have made the popular list but we don't have the data beyond 2011.
- Item 6452-TORTILLA/TOSTADA CHIPS was made in popular list in 2008, however it was the most expensive item in salty snack in that year, that could be the reason why it could not make it the top 3 list in subsequent years.

## 3. How is the popularity affected over the years? 
Answers: 
- The popular items usual will be popular at least 2 years in sequence.  I also did the same analysis for 2001-2007, the top 3 popular list has changed in within these 3 categories over 7 years.  
- In 2008-2011, the top 3 popular list has 3 categories, and within these 3 categories, two of items (7858 and 7881) seem to repeatedly making the list.
- Price seems to play a major role.  If the price is much higher than the average of the same category, it will be dropped out the popular list.

### Continue to clean the data for Nearest Neighbors regression

In [8]:
print store_df.dtypes, "\n"
print store_df.columns

IRI_KEY            int64
WEEK               int64
SY                 int64
GE                 int64
VEND               int64
ITEM               int64
UNITS              int64
DOLLARS          float64
F                 object
D                  int64
PR                 int64
DOLLARS_2        float64
UNITS_2            int64
DOLLARS_2_SUM    float64
UNITS_2_SUM        int64
DOLLARS_AVG      float64
DOLLARS_SUM      float64
UNITS_SUM          int64
dtype: object 

Index([u'IRI_KEY', u'WEEK', u'SY', u'GE', u'VEND', u'ITEM', u'UNITS',
       u'DOLLARS', u'F', u'D', u'PR', u'DOLLARS_2', u'UNITS_2',
       u'DOLLARS_2_SUM', u'UNITS_2_SUM', u'DOLLARS_AVG', u'DOLLARS_SUM',
       u'UNITS_SUM'],
      dtype='object')


In [9]:
feature_columns = [u'IRI_KEY', u'WEEK', 'VEND', u'ITEM', u'UNITS', u'DOLLARS', \
                  u'DOLLARS_2', u'UNITS_2', u'DOLLARS_2_SUM', u'UNITS_2_SUM', u'DOLLARS_AVG']
store2_df = store_df[feature_columns]
print "store2_df.shape=", store2_df.shape

store2_df.shape= (6806768, 11)


In [10]:
store2_df.describe().T

count          mean           std        min  \
IRI_KEY        6806768.0  3.827444e+06  3.569401e+06  235547.00   
WEEK           6806768.0  1.584902e+03  6.035125e+01    1479.00   
VEND           6806768.0  3.976685e+04  2.228631e+04      16.00   
ITEM           6806768.0  1.304332e+04  1.678290e+04       1.00   
UNITS          6806768.0  3.937769e+00  5.358526e+00       1.00   
DOLLARS        6806768.0  6.534841e+00  7.800154e+00       0.01   
DOLLARS_2      6806768.0  1.966894e+00  1.143826e+00       0.01   
UNITS_2        6806768.0  1.000000e+00  0.000000e+00       1.00   
DOLLARS_2_SUM  6806768.0  1.208654e+04  1.366368e+04       0.09   
UNITS_2_SUM    6806768.0  6.020244e+03  5.230971e+03       1.00   
DOLLARS_AVG    6806768.0  1.966894e+00  1.059532e+00       0.09   

                         25%           50%           75%           max  
IRI_KEY        664475.000000  1.110947e+06  8.014600e+06  8.057509e+06  
WEEK             1532.000000  1.587000e+03  1.638000e+03  1.686000e+03  
VEND            28400.000000  2.840000e+04  4.163300e+04  9.999800e+04  
ITEM             6399.000000  8.102000e+03  1.104400e+04  9.990300e+04  
UNITS               1.000000  2.000000e+00  5.000000e+00  1.185000e+03  
DOLLARS             2.190000  3.980000e+00  7.930000e+00  5.089800e+02  
DOLLARS_2           0.990000  1.500000e+00  2.790000e+00  1.999000e+01  
UNITS_2             1.000000  1.000000e+00  1.000000e+00  1.000000e+00  
DOLLARS_2_SUM    2489.703528  7.074205e+03  1.603352e+04  6.819172e+04  
UNITS_2_SUM      1541.000000  4.446000e+03  9.621000e+03  1.992500e+04  
DOLLARS_AVG         0.997757  1.719734e+00  2.775861e+00  1.773170e+01

### Table mergings

In [11]:
import numpy as np
store2_np = np.array(store2_df)

In [12]:
# indexes = np.where(store2_np[:,4] == 0)
# print len(indexes[0])
#store2_np.shape

In [13]:
%sql postgresql://sharknado:sharknado@sharknado-dse.ceg3hdkdq8l0.us-east-1.rds.amazonaws.com/sharknado
result = %sql select * from week_season
week_season_df = result.DataFrame()

week_season_df = week_season_df[['iri_week', 'season_numeric']]
week_season_df.columns = ['WEEK', 'SEASON']

522 rows affected.


In [14]:
%sql postgresql://sharknado:sharknado@sharknado-dse.ceg3hdkdq8l0.us-east-1.rds.amazonaws.com/sharknado
result = %sql select distinct item,vend,l2,l3 from products
product_df = result.DataFrame()

35547 rows affected.


In [15]:
print "product_df.shape", product_df.shape
product_df.describe().T
product_df.columns = ['ITEM', 'VEND', 'L2', 'L3']

product_df['ITEM'] = product_df['ITEM'].astype(int)
product_df['VEND'] = product_df['VEND'].astype(int)

print product_df.dtypes

product_df.shape (35547, 4)
ITEM     int64
VEND     int64
L2      object
L3      object
dtype: object


In [16]:
store_product_df = pd.merge(left=store2_df, right=product_df, on=['ITEM','VEND'], how='left')
store_product_df = pd.merge(left=store_product_df, right=week_season_df, on=['WEEK'], how='left')

print "store_product_df.shape=", store_product_df.shape

store_product_df.shape= (7512405, 14)


In [17]:
store_product_df.head(1)

IRI_KEY  WEEK   VEND  ITEM  UNITS  DOLLARS  DOLLARS_2  UNITS_2  \
0   681530  1479  28400  4874      2     1.98       0.99        1   

   DOLLARS_2_SUM  UNITS_2_SUM  DOLLARS_AVG            L2           L3  SEASON  
0    7111.326874         7292     0.975223  POTATO CHIPS  PEPSICO INC       4

### Feature selection

In [18]:
L2_type = [u'POTATO CHIPS', u'OTHER SALTED SNACKS (NO NUTS)',
       u'TORTILLA/TOSTADA CHIPS', u'PRETZELS', u'CHEESE SNACKS',
       u'READY-TO-EAT POPCORN/CARAMEL COR',
       u'CORN SNACKS (NO TORTILLA CHIPS)', u'PORK RINDS']

def changeEnumValue(x, l):
    for idx, val in enumerate(l):
        #print x, val
        if x==val:
            return idx
            
    if type(x) == int:
        return x
    
feature_engineering_columns = [u'SEASON', 'ITEM', u'DOLLARS_2', u'L2']
store_product_df2 = store_product_df[feature_engineering_columns]
store_product_df2.L2 = store_product_df2.L2.apply(changeEnumValue, args=[L2_type])

store_product_df2.DOLLARS_2 = store_product_df2.DOLLARS_2.apply(lambda x:'%.2f'%x)
store_product_df2.DOLLARS_2 = store_product_df2.DOLLARS_2.astype(float)

store_product_df2.head(5)

SEASON  ITEM  DOLLARS_2  L2
0       4  4874       0.99   0
1       4  5189       0.99   0
2       4  6389       0.99   0
3       4  6374       3.00   0
4       4  6390       0.99   0

## 4. Extra Credit:

## Nearest Neighbors regression
- To predict the price of an item of its category durning certain season (winter, spring, summer, fall)

In [19]:
from sklearn.cross_validation import train_test_split
store_product_np = np.array(store_product_df2)

X = store_product_np[:,[0,1,3]]
y = store_product_np[:,2]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)
print "shapes", X_train.shape, X_test.shape, y_train.shape, y_test.shape

shapes (5258683, 3) (2253722, 3) (5258683,) (2253722,)


In [20]:
# X_train = store_product_np[:100000,[0,1,3]]
# y_train = store_product_np[:100000,2]
# X_test = store_product_np[100000:200000,[0,1,3]]
# y_test = store_product_np[100000:200000,2]

In [21]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn import neighbors

np.random.seed(0)
n_neighbors = 5

for i, weights in enumerate(['distance']):
    knn = neighbors.KNeighborsRegressor(n_neighbors, weights=weights).fit(X_train, y_train) #.ravel()
    scores = knn.score(X_test, y_test)
    print 'Accuracy is:','%.2f'%(scores*100),'%'

Accuracy is: 83.01 %


## Conclusion:  
- Tell the season, item, salty-snack category that the user would like to purchase, this KNeighborsRegressor model can predict that item's price.   And the accuracy is up to 83%